mautic/mautic
# Marketplace v1

| Q                                      | A
| -------------------------------------- | ---
| Branch?                                | "features"
| Bug fix?                               | no
| New feature?                           | yes
| Deprecations?                          | no
| BC breaks?                             | no
| Automated tests included?              | yes
| Related user documentation PR URL      | https://github.com/mautic/mautic-documentation/pull/185
| Related developer documentation PR URL | https://github.com/mautic/developer-documentation/pull/196
| Issue(s) addressed                     | /

#### Description:

The code for this read-only version of the Mautic Marketplace is mostly taken from my previous PR at https://github.com/mautic/mautic/pull/8257. I moved the MauticMarketplaceBundle to the core bundles instead of a plugin and removed my experimentation with Composer installations.

The reason this is read-only first is that we are waiting for Mautic 4 directory structure and Composer 2 to make the installations possible.

##### The list view
The data are directly loaded from Packagist.org API. They do not allow to order by specific values. It would be nice to order by some though. The search box works.
<img width="1224" alt="Screen Shot 2021-02-06 at 20 30 40" src="https://user-images.githubusercontent.com/1235442/107127983-5f7a7a00-68ba-11eb-8e39-7c5dfcb192f8.png">

##### The detail view
This is the view I'd like some help from a UI/UX person.
<img width="1227" alt="Screen Shot 2021-02-06 at 20 31 08" src="https://user-images.githubusercontent.com/1235442/107127988-6acda580-68ba-11eb-9dde-600cec7fee56.png">

##### Permissions
Only the view and create (=install) permissions are implemented. The create permissions only show/hide the install button for now. This button does not work anyway and just shows a modal that the one click install is coming soon. But the permissions are prepared there. The edit permission can be used once the user will be able to configure which stability (stable, beta, dev, ...) or certain version they want installed for each plugin. And the delete permission is reserved for deleting a plugin installed via the marketplace..
<img width="686" alt="Screen Shot 2021-02-06 at 20 31 31" src="https://user-images.githubusercontent.com/1235442/107127993-74570d80-68ba-11eb-9bd3-094ed96dd045.png">

##### Feature flag to enable/disable the marketplace
You can disable the Mautic Marketplace from the app/config/local.php file. It is enabled by default. This is useful for Mautic providers who cannot enable the Marketplace due to security concerns.

Paste `'marketplace_enabled' => false,` to the local.php file to disable it. It will hide:
- The menu item
- The permission settings
- The marketplace routs will return 404 if accessed directly via URL

#### ToDo
- [x] Functional test of the `mautic:marketplace:list` command
- [x] Functional test of the list page
- [x] Functional test of the detail page
- [x] User docs (note: it's read-only for now!)
- [x] Dev docs (how to create plugins and make them visible in the Marketplace)

#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2. Go to the Marketplace (see the attached user docs) and click around. There should be no error and all the info presented should make sense.
3. The marketplace is read-only now. The only functionality other than viewing the info is the search on the list view.
4. There are many links to another websites in the UI. Click on them randomly and make sure it leads you to the place you'd expect.
5. If you test on local, test also the CLI command.
